### PR TITLE
Typecheck configuration p3

### PR DESCRIPTION
**Pull request by smohsinali**

### CODE REVIEW

File: ConfigSpace/configuration_space.py [hunk 1]
```diff
@@ -48,34 +50,35 @@ class ConfigurationSpace(object):
...
-        self._children = OrderedDict()
-        self._parents = OrderedDict()
+        self._children = OrderedDict()   # type: OrderedDict[str, OrderedDict[str, Union[None, AbstractCondition]]]
+        self._parents = OrderedDict()   # type: OrderedDict[str, OrderedDict[str, Union[None, AbstractCondition]]]
 
         # changing this to a normal dict will break sampling because there is
         #  no guarantee that the parent of a condition was evaluated before
-        self._conditionsals = OrderedDict()
-        self.forbidden_clauses = []
+        self._conditionsals = OrderedDict()   # type: OrderedDict[str, str]
```
Comment (mfeurer) [inline ConfigSpace/configuration_space.py]: Line 169 suggests that this maps for str to Condition or Hyperparameter.

File: ConfigSpace/configuration_space.py [hunk 2]
```diff
@@ -113,9 +116,10 @@ def add_hyperparameter(self, hyperparameter):
         self._check_default_configuration()
         self._sort_hyperparameters()
 
+        # todo: check it returns input as aoutput?
         return hyperparameter
 
-    def add_condition(self, condition):
+    def add_condition(self, condition: ConditionComponent) -> ConditionComponent:
```
Comment (mfeurer) [inline ConfigSpace/configuration_space.py]: Shouldn't this accept and return an abstract condition?

File: ConfigSpace/configuration_space.py [hunk 3]
```diff
@@ -222,7 +226,7 @@ def _sort_hyperparameters(self):
                 else:
                     to_visit.appendleft(current)
 
-        by_level = defaultdict(list)
+        by_level = defaultdict(list)  # type: Dict['int', List[str]]
```
Comment (mfeurer) [inline ConfigSpace/configuration_space.py]: Why not `#type: defaultdict[int, List[str]]`?

File: ConfigSpace/configuration_space.py [hunk 4]
```diff
@@ -285,8 +289,8 @@ def add_forbidden_clause(self, clause):
     #     HPOlibConfigSpace.nx.draw(self._dg, pos, with_labels=True)
     #     plt.savefig('nx_test.png')
 
-    def add_configuration_space(self, prefix, configuration_space,
-                                delimiter=":", parent_hyperparameter=None):
+    def add_configuration_space(self, prefix: str, configuration_space: ConfigSpace,
```
Comment (mfeurer) [inline ConfigSpace/configuration_space.py]: It should accept and return `ConfigurationSpace`, right?

File: ConfigSpace/configuration_space.py [hunk 5]
```diff
@@ -396,16 +401,16 @@ def get_conditions(self):
...
+        parents = [] # type: List[Hyperparameter]
         for condition in conditions:
             parents.extend(condition.get_children())
         return parents
 
-    def get_child_conditions_of(self, name):
+    def get_child_conditions_of(self, name: Union[str, Hyperparameter]) -> List[AbstractCondition]:
         if isinstance(name, Hyperparameter):
-            name = name.name
+            name = name.name  # type: ignore
```
Comment (mfeurer) [inline ConfigSpace/configuration_space.py]: What about `# type: str`?

File: ConfigSpace/configuration_space.py [hunk 6]
```diff
@@ -437,41 +442,42 @@ def get_parents_of(self, name):
...
+        parents = [] # type: List[Hyperparameter]
         for condition in conditions:
             parents.extend(condition.get_parents())
         return parents
 
-    def get_parent_conditions_of(self, name):
+    def get_parent_conditions_of(self, name: Union[str, Hyperparameter]) -> List[AbstractCondition]:
         if isinstance(name, Hyperparameter):
-            name = name.name
+            name = name.name  # type: ignore
```
Comment (mfeurer) [inline ConfigSpace/configuration_space.py]: What about `# type: str`?

File: ConfigSpace/configuration_space.py [hunk 7]
```diff
@@ -437,41 +442,42 @@ def get_parents_of(self, name):
...
 
-    def _get_parent_conditions_of(self, name):
+    def _get_parent_conditions_of(self, name: str) -> List[AbstractCondition]:
         parents = self._parents[name]
         conditions = [parents[parent_name] for parent_name in parents
                       if parent_name != "__HPOlib_configuration_space_root__"]
         return conditions
-
-    def get_all_unconditional_hyperparameters(self):
+    # todo: find out if hyperparameter obj or only its name is returned?
```
Comment (mfeurer) [inline ConfigSpace/configuration_space.py]: Please check this with the debugger and the unit tests.

File: ConfigSpace/configuration_space.py [hunk 8]
```diff
@@ -636,10 +638,10 @@ def sample_configuration(self, size=1):
                 vector[:, i] = hyperparameter._sample(self.random, missing)
 
             for i in range(missing):
-                inactive = set()
+                inactive = set()  # type: Set['str']
                 visited = set()
                 visited.update(self.get_all_unconditional_hyperparameters())
-                to_visit = deque()
+                to_visit = deque()  # type: ignore
```
Comment (mfeurer) [inline ConfigSpace/configuration_space.py]: Why do you have to ignore this?

File: ConfigSpace/configuration_space.py [hunk 9]
```diff
@@ -702,15 +703,17 @@ def sample_configuration(self, size=1):
...
 
 class Configuration(object):
     # TODO add a method to eliminate inactive hyperparameters from a
     # configuration
-    def __init__(self, configuration_space, values=None, vector=None,
-                 allow_inactive_with_values=False, origin=None):
+    # todo check types of vector and origin
+    def __init__(self, configuration_space: ConfigurationSpace, values: Union[None, Dict[str, Hyperparameter]] = None,
+                 vector: Union[None, np.ndarray]=None, allow_inactive_with_values: bool=False, origin: Any=None)\
+            -> None:
```
Comment (mfeurer) [inline ConfigSpace/configuration_space.py]: I think value needs to be `Union[None, Dict[str, Union[str, float, int]]]`.

File: ConfigSpace/configuration_space.py [hunk 10]
```diff
@@ -741,7 +744,7 @@ def __init__(self, configuration_space, values=None, vector=None,
         self._query_values = False
         self._num_hyperparameters = len(self.configuration_space._hyperparameters)
         self.origin = origin
-        self._keys = None
+        self._keys = None  # type: ignore
```
Comment (mfeurer) [inline ConfigSpace/configuration_space.py]: Why this?

File: ConfigSpace/configuration_space.py [hunk 11]
```diff
@@ -750,7 +753,7 @@ def __init__(self, configuration_space, values=None, vector=None,
             # Using cs._hyperparameters to iterate makes sure that the
             # hyperparameters in the configuration are sorted in the same way as
             # they are sorted in the configuration space
-            self._values = dict()
+            self._values = dict()  # type: ignore
```
Comment (mfeurer) [inline ConfigSpace/configuration_space.py]: Why this?

File: ConfigSpace/configuration_space.py [hunk 12]
```diff
@@ -808,19 +811,19 @@ def __getitem__(self, item):
             value = float(repr(value))
         self._values[item] = value
         return self._values[item]
-
-    def get(self, item, default=None):
+    # todo: find return type of get
```
Comment (mfeurer) [inline ConfigSpace/configuration_space.py]: I think this is correct.

File: ConfigSpace/configuration_space.py [hunk 13]
```diff
@@ -808,19 +811,19 @@ def __getitem__(self, item):
...
+    def get(self, item: str, default: Union[None, Any]=None) -> Union[None, Any]:
         try:
             return self[item]
         except:
             return default
-
-    def __contains__(self, item):
+    # todo: recheck
```
Comment (mfeurer) [inline ConfigSpace/configuration_space.py]: This should be `bool` I think.

File: ConfigSpace/configuration_space.py [hunk 14]
```diff
@@ -868,21 +871,21 @@ def __repr__(self):
...
+    def keys(self) -> List[str]:
         # Cache the keys to speed up the process of retrieving the keys
         if self._keys is None:
             self._keys = list(self.configuration_space._hyperparameters.keys())
         return self._keys
 
-
-    def get_dictionary(self):
+    # todo: recheck
+    def get_dictionary(self) -> Any:
```
Comment (mfeurer) [inline ConfigSpace/configuration_space.py]: This returns a dictionary ;)

**Comment by mfeurer** [comment]
Could you please also make the sorting in line 648 deterministic to prevent stochastic failure of unit tests?

File: ConfigSpace/configuration_space.py (hunk 1 above)
Comment (smohsinali) [inline ConfigSpace/configuration_space.py]: line 169 is 
**self._conditionsals[child_node] = child_node**
from here it appears that key and value are equal. debugger also says the same

File: ConfigSpace/configuration_space.py (hunk 2 above)
Comment (smohsinali) [inline ConfigSpace/configuration_space.py]: the first line of this function raises error if input is not of type conditioncomponent
        **if not isinstance(condition, ConditionComponent):
            raise TypeError("The method add_condition must be called "
                            "with an instance of "
                            "HPOlibConfigSpace.condition.ConditionComponent.")**

Comment (smohsinali) [inline ConfigSpace/configuration_space.py]: return type changed to abstractcondition

File: ConfigSpace/configuration_space.py (hunk 3 above)
Comment (smohsinali) [inline ConfigSpace/configuration_space.py]: changed

File: ConfigSpace/configuration_space.py (hunk 4 above)
Comment (smohsinali) [inline ConfigSpace/configuration_space.py]: fixed

File: ConfigSpace/configuration_space.py (hunk 5 above)
Comment (smohsinali) [inline ConfigSpace/configuration_space.py]: i thought so too but mypy gives following error which doesnt actually make any sense
**configuration_space.py:413: error: Some element of union has no attribute "name"**

File: ConfigSpace/configuration_space.py (hunk 6 above)
Comment (smohsinali) [inline ConfigSpace/configuration_space.py]: same as above

File: ConfigSpace/configuration_space.py (hunk 7 above)
Comment (smohsinali) [inline ConfigSpace/configuration_space.py]: done...it returns only list of names and not full objects

File: ConfigSpace/configuration_space.py (hunk 8 above)
Comment (smohsinali) [inline ConfigSpace/configuration_space.py]: fixed

File: ConfigSpace/configuration_space.py (hunk 9 above)
Comment (smohsinali) [inline ConfigSpace/configuration_space.py]: changed

File: ConfigSpace/configuration_space.py (hunk 10 above)
Comment (smohsinali) [inline ConfigSpace/configuration_space.py]: fixed

File: ConfigSpace/configuration_space.py (hunk 1 above)
Comment (mfeurer) [inline ConfigSpace/configuration_space.py]: Hm, let's discuss this later.

File: ConfigSpace/configuration_space.py (hunk 2 above)
Comment (mfeurer) [inline ConfigSpace/configuration_space.py]: You're right, ConditionComponent is a parent class of AbstractCondition; accepting and returning a ConditionComponent is totally correct.

File: ConfigSpace/configuration_space.py (hunk 7 above)
Comment (mfeurer) [inline ConfigSpace/configuration_space.py]: Okay, then please remove the comment.

File: ConfigSpace/configuration_space.py (hunk 11 above)
Comment (smohsinali) [inline ConfigSpace/configuration_space.py]: changed

File: ConfigSpace/configuration_space.py (hunk 13 above)
Comment (smohsinali) [inline ConfigSpace/configuration_space.py]: return type changed to bool

File: ConfigSpace/configuration_space.py (hunk 14 above)
Comment (smohsinali) [inline ConfigSpace/configuration_space.py]: changed to **Dict[str, Union[str, float, int]]**

File: ConfigSpace/configuration_space.py (hunk 1 above)
Comment (smohsinali) [inline ConfigSpace/configuration_space.py]: make it a set

File: ConfigSpace/configuration_space.py [hunk 15]
```diff
@@ -119,7 +119,7 @@ def add_hyperparameter(self, hyperparameter: Hyperparameter) -> Hyperparameter:
         # todo: check it returns input as aoutput?
         return hyperparameter
 
-    def add_condition(self, condition: ConditionComponent) -> ConditionComponent:
+    def add_condition(self, condition: ConditionComponent) -> AbstractCondition:
```
Comment (mfeurer) [inline ConfigSpace/configuration_space.py]: This was correct before, please revert.

File: ConfigSpace/configuration_space.py (hunk 1 above)
Comment (mfeurer) [inline ConfigSpace/configuration_space.py]: This is still open, right?

**Comment by mfeurer** [comment]
Please also remove all TODOs that you added regarding types; together with the two changes to go, this will be ready to merge.

File: ConfigSpace/configuration_space.py (hunk 15 above)
Comment (smohsinali) [inline ConfigSpace/configuration_space.py]: reverted

**Pull request merged**